include/ringmesh/basic/frame.h

```diff
@@ -0,0 +1,188 @@
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL ASGA BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *     http://www.ring-team.org
+ *
+ *     RING Project
+ *     Ecole Nationale Superieure de Geologie - GeoRessources
+ *     2 Rue du Doyen Marcel Roubault - TSA 70605
+ *     54518 VANDOEUVRE-LES-NANCY
+ *     FRANCE
+ */
+
+#pragma once
+
+#include <array>
+
+#include <ringmesh/basic/common.h>
+
+/*!
+ * @file Basic classes for representing frames
+ * @author Pierre Anquez
+ */
+
+namespace RINGMesh
+{
+    namespace Geometry
+    {
+        struct Plane;
+        FORWARD_DECLARATION_DIMENSION_STRUCT( Line );
+        ALIAS_2D( Line );
+    }
+}
+
+namespace RINGMesh
+{
+    template < index_t DIMENSION >
+    class RINGMESH_API FrameBase
+    {
+    public:
+        virtual ~FrameBase() = default;
+
+        const vecn< DIMENSION >& operator[]( index_t coord ) const
+        {
+            ringmesh_assert( coord < DIMENSION );
+            return axis_[coord];
+        }
+
+        vecn< DIMENSION >& operator[]( index_t coord )
+        {
+            ringmesh_assert( coord < DIMENSION );
+            return axis_[coord];
+        }
+
+    protected:
+        // Default Frame aligned on space axis
+        FrameBase()
+        {
+            for( auto coord : RINGMesh::range( DIMENSION ) )
+            {
+                axis_[coord][coord] = 1.;
+            }
+        }
+
+    private:
+        std::array< vecn< DIMENSION >, DIMENSION > axis_;
+    };
+    ALIAS_2D_AND_3D( FrameBase );
+
+    template < index_t DIMENSION >
+    class RINGMESH_API Frame : public FrameBase< DIMENSION >
+    {
+    public:
+        Frame() = default;
+    };
+    ALIAS_2D_AND_3D( Frame );
+
+    template <>
+    class RINGMESH_API Frame< 2 > : public FrameBase< 2 >
+    {
+    public:
+        Frame() = default;
+
+        Frame( vec2 x, vec2 y )
+        {
+            ( *this )[0] = std::move( x );
+            ( *this )[1] = std::move( y );
+        }
+    };
+
+    template <>
+    class RINGMESH_API Frame< 3 > : public FrameBase< 3 >
+    {
+    public:
+        Frame() = default;
+
+        Frame( vec3 x, vec3 y, vec3 z )
+        {
+            ( *this )[0] = std::move( x );
```

Comment:
Do you think we should move the  common two lines (duplication with Frame<2> constructor) into a constructor of FrameBase?

Reply:
Not necessary... I think it is clearer like this. What do you think?

Reply:
It is clearer but it is code dupplication